### PR TITLE
feat: [CDS-48056]: Rolling Deployment and Rollback For TAS

### PR DESCRIPTION
**Pull request by piyush-bhuwalka**
## Describe your changes

Rolling Deploy and Rolling Rollback Steps implemented successfully. And Manager side changes for Tas Mapping Step.
https://harness.atlassian.net/wiki/spaces/CDP/pages/21211480091/PCF+NG+-+YAML
https://harness.atlassian.net/wiki/spaces/PD/pages/21210531200/PCF+NG+-+Technical+Proposal#Use-case4---Rolling-Deployment
https://docs.pivotal.io/application-service/2-7/devguide/deploy-apps/rolling-deploy.html

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/41907)

### CODE REVIEW

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by piyush-bhuwalka** [comment]
trigger messagecheck

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by piyush-bhuwalka** [comment]
trigger messagecheck

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

File: 125-cd-nextgen/src/main/java/io/harness/cdng/creator/CDNGPlanCreatorProvider.java [hunk 1]
```diff
@@ -1096,6 +1108,29 @@ public List<StepInfo> getStepInfo() {
...
+    StepInfo tasRollingRollback =
+        StepInfo.newBuilder()
+            .setName("Rolling Rollback")
+            .setType(StepSpecTypeConstants.TAS_ROLLING_ROLLBACK)
+            .setStepMetaData(StepMetaData.newBuilder().addCategory(TAS).setFolderPath(TAS).build())
+            .setFeatureFlag(FeatureName.CDS_TAS_NG.name())
+            .build();
+
+    StepInfo routeMapping = StepInfo.newBuilder()
+                                .setName("Route Mapping")
```
Comment (sourabh-awashti) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/creator/CDNGPlanCreatorProvider.java:1128]: change route mapping variable to tas route mapping

File: 125-cd-nextgen/src/main/java/io/harness/cdng/creator/CDCreatorUtils.java [hunk 2]
```diff
@@ -53,6 +53,7 @@ public Set<String> getSupportedStepsV2() {
         StepSpecTypeConstants.TAS_APP_RESIZE, StepSpecTypeConstants.TAS_SWAP_ROUTES, StepSpecTypeConstants.TAS_ROLLBACK,
         StepSpecTypeConstants.SWAP_ROLLBACK, StepSpecTypeConstants.ELASTIGROUP_BG_STAGE_SETUP,
         StepSpecTypeConstants.ELASTIGROUP_SWAP_ROUTE, StepSpecTypeConstants.ASG_ROLLING_DEPLOY,
-        StepSpecTypeConstants.ASG_ROLLING_ROLLBACK);
+        StepSpecTypeConstants.ASG_ROLLING_ROLLBACK, StepSpecTypeConstants.TAS_ROLLING_DEPLOY,
+        StepSpecTypeConstants.TAS_ROLLING_ROLLBACK, StepSpecTypeConstants.ROUTE_MAPPING);
   }
```
Comment (sourabh-awashti) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/creator/CDCreatorUtils.java:58]: TAS_ROUTE_MAPPING

File: 125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/steps/TasRouteMappingStepPlanCreator.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+import com.google.common.collect.Sets;
+import java.util.Set;
+
+@OwnedBy(HarnessTeam.CDP)
+public class TasRouteMappingStepPlanCreator extends CDPMSStepPlanCreatorV2<TasRouteMappingStepNode> {
+  @Override
+  public Set<String> getSupportedStepTypes() {
+    return Sets.newHashSet(ROUTE_MAPPING);
+  }
```
Comment (sourabh-awashti) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/steps/TasRouteMappingStepPlanCreator.java:27]: check with other step types

File: 125-cd-nextgen/src/main/java/io/harness/cdng/tas/TasRollingDeployStep.java [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+  }
+
+  @Override
+  public TaskChainResponse executeTasTask(ManifestOutcome tasManifestOutcome, Ambiance ambiance,
+      StepElementParameters stepParameters, TasExecutionPassThroughData executionPassThroughData,
+      boolean shouldOpenFetchFilesLogStream, UnitProgressData unitProgressData) {
+    TasRollingDeployStepParameters tasRollingDeployStepParameters =
+        (TasRollingDeployStepParameters) stepParameters.getSpec();
+    ArtifactOutcome artifactOutcome = cdStepHelper.resolveArtifactsOutcome(ambiance).orElseThrow(
+        () -> new InvalidArgumentsException(Pair.of("artifacts", "Primary artifact is required for PCF")));
```
Comment (sourabh-awashti) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/tas/TasRollingDeployStep.java:216]: change to TAS

File: 125-cd-nextgen/src/main/java/io/harness/cdng/tas/TasRollingDeployStep.java [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+
+      return StepResponse.builder()
+          .status(Status.SUCCEEDED)
+          .unitProgressList(response.getUnitProgressData().getUnitProgresses())
+          .stepOutcome(stepOutcome)
+          .stepOutcome(StepResponse.StepOutcome.builder()
+                           .outcome(tasRollingDeployOutcome)
+                           .name(OutcomeExpressionConstants.TAS_ROLLING_DEPLOY_OUTCOME)
+                           .group(StepCategory.STAGE.name())
+                           .build())
```
Comment (sourabh-awashti) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/tas/TasRollingDeployStep.java:191]: replace stage with step

File: 400-rest/src/main/java/software/wings/helpers/ext/pcf/CfDeploymentManagerUnsupported.java [hunk 6]
```diff
@@ -193,4 +193,16 @@ public void runPcfPluginScript(CfRunPluginScriptRequestData requestData, LogCall
       throws PivotalClientApiException {
     throw new PivotalClientApiException("PCF operations not supported by this API.");
   }
+
+  @Override
+  public ApplicationDetail createRollingApplicationWithSteadyStateCheck(CfCreateApplicationRequestData requestData,
+      LogCallback executionLogCallback) throws PivotalClientApiException, InterruptedException {
+    throw new PivotalClientApiException("PCF operations not supported by this API.");
```
Comment (sourabh-awashti) [inline 400-rest/src/main/java/software/wings/helpers/ext/pcf/CfDeploymentManagerUnsupported.java:200]: replace with TAS

File: 930-delegate-tasks/src/main/java/io/harness/delegate/pcf/CfRollingRollbackCommandTaskHandlerNG.java [hunk 7]
```diff
@@ -0,0 +1,608 @@
...
+    //        ArrayList<Map<String, Object>> allProcesses = (ArrayList<Map<String, Object>>) processes;
+    //        for (Map<String, Object> process : allProcesses) {
+    //          Object p = process.get(PROCESSES_TYPE_MANIFEST_YML_ELEMENT);
+    //          if ((p instanceof String) && (p.toString().equals(WEB_PROCESS_TYPE_MANIFEST_YML_ELEMENT))) {
+    //            process.put(INSTANCE_MANIFEST_YML_ELEMENT, 0);
+    //          }
+    //        }
+    //      }
+    //    }
+    // Update routes.
```
Comment (sourabh-awashti) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/pcf/CfRollingRollbackCommandTaskHandlerNG.java:417]: remove

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/cf/CfCommandTaskHelperNG.java [hunk 8]
```diff
@@ -651,6 +651,11 @@ public ApplicationDetail getNewlyCreatedApplication(CfRequestConfig cfRequestCon
     return cfDeploymentManager.getApplicationByName(cfRequestConfig);
   }
 
+  public ApplicationDetail getApplicationDetails(
+      CfRequestConfig cfRequestConfig, CfDeploymentManager cfDeploymentManager) throws PivotalClientApiException {
+    return cfDeploymentManager.getApplicationByName(cfRequestConfig);
+  }
+
```
Comment (sourabh-awashti) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/cf/CfCommandTaskHelperNG.java:658]: remove this function

File: 960-api-services/src/main/java/io/harness/pcf/CfSdkClient.java [hunk 9]
```diff
@@ -36,7 +36,6 @@ public interface CfSdkClient {
    */
   List<OrganizationSummary> getOrganizations(CfRequestConfig pcfRequestConfig)
       throws PivotalClientApiException, InterruptedException;
-
```
Comment (sourabh-awashti) [inline 960-api-services/src/main/java/io/harness/pcf/CfSdkClient.java:39]: add line

File: 960-api-services/src/main/java/io/harness/pcf/cfsdk/CfSdkClientImpl.java [hunk 10]
```diff
@@ -120,7 +120,6 @@ public List<OrganizationSummary> getOrganizations(CfRequestConfig pcfRequestConf
     }
     return organizations;
   }
-
```
Comment (sourabh-awashti) [inline 960-api-services/src/main/java/io/harness/pcf/cfsdk/CfSdkClientImpl.java:123]: revert

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/service/beans/ServiceDefinitionType.java [hunk 11]
```diff
@@ -71,7 +71,7 @@ public enum ServiceDefinitionType {
   @JsonProperty(ServiceSpecType.TAS)
   TAS(ServiceSpecType.TAS,
       Lists.newArrayList(ExecutionStrategyType.BASIC, ExecutionStrategyType.CANARY, ExecutionStrategyType.BLUE_GREEN,
-          ExecutionStrategyType.DEFAULT),
+          ExecutionStrategyType.DEFAULT, ExecutionStrategyType.ROLLING),
       ServiceSpecType.TAS),
```
Comment (sourabh-awashti) [inline 127-cd-nextgen-entities/src/main/java/io/harness/cdng/service/beans/ServiceDefinitionType.java:75]: add rolling with verify yaml

File: 125-cd-nextgen/src/main/java/io/harness/cdng/tas/TasSwapRoutesStep.java [hunk 12]
```diff
@@ -77,7 +77,7 @@
 
 @OwnedBy(HarnessTeam.CDP)
 @Slf4j
-public class TasSwapRoutesStep extends CdTaskExecutable<CfCommandResponseNG> {
+public class  TasSwapRoutesStep extends CdTaskExecutable<CfCommandResponseNG> {
   public static final StepType STEP_TYPE = StepType.newBuilder()
```
Comment (sourabh-awashti) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/tas/TasSwapRoutesStep.java:81]: revert

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Review by ivan-83 — APPROVED** [review]
:lgtm:

> Reviewed 59 of 82 files at r1, 24 of 24 files at r2, 11 of 11 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41907#-:-NLfClzvE9MUliLNRnOH:bnfp4nl)* status: 84 file(s) to review. 1 of 2 approvals obtained.

**Review by anilchowdhury-harness — DISMISSED** [review]
:lgtm:

> Reviewed 51 of 82 files at r1, 22 of 24 files at r2, 11 of 11 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41907#-:-NLfKiaY9sahBGlvPyyp:bnfp4nl)* status: :shipit: complete! ✓

**Review by yogesh-chauhan — APPROVED** [review]

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 48 of 82 files at r1, 19 of 24 files at r2, 10 of 11 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41907)* status: 2 file(s) to review. 1 of 2 approvals obtained.

**Comment by piyush-bhuwalka** [comment]
trigger ti1

**Review by anilchowdhury-harness — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41907#-:-NLg2Hl53qXJdwlVbcDf:bnfp4nl)* status: :shipit: complete! ✓

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Pull request merged**